[RFC 0153] Non-legacy boot NixOS (#154)
NixOS/rfcs: rfcs/0153-uefi-only.md

---
feature: non_legacy_boot
start-date: 2023-06-16
author: Ryan Lahfa
co-authors: (find a buddy later to help out with the RFC)
shepherd-team: (names, to be nominated and accepted by RFC steering committee)
shepherd-leader: (name to be appointed by RFC steering committee)
related-issues: (will contain links to implementation PRs)
---

# Summary
[summary]: #summary

NixOS will have first-class support for UEFI

> and uses it as a default boot environment, for supported architectures,

Reviewer: ```suggestion
and uses it as a default boot environment, on supported machines,
```

UEFI does not "support" entire architectures in any meaningful sense.  It "supports" MIPS and yet there is not a single UEFI MIPS machine in existence!  Talking about "UEFI-supported architectures" is extremely misleading.

even in situations where only BIOS Boot Specification's legacy boot is available,

> via a dual-stage payload consisting of a polyfill bootloader/firmware and a standard

Reviewer: Please no.  If you want these Rube Goldberg contraptions on your machine, great.  Don't force them on everybody else.

Reviewer: I stopped reading at this point.

UEFI bootloader.

To achieve this, it will downgrade the GRUB privileged position
in the project and offer it as a "best effort" basis alternative bootloader.

# Motivation
[motivation]: #motivation

Legacy boot is defined by the BIOS Boot specification, written in 1996: https://www.scs.stanford.edu/nyu/04fa/lab/specsbbs101.pdf.
Nowadays, computers are defaulting to UEFI more and more for the extended features provided (e.g. SecureBoot, native network boot, etc.).

Nevertheless, many legacy boots machines or machines that does not have support for UEFI are used with NixOS: Single Board Computers for example
on other architectures.

Let's put aside non-legacy boot payloads such as [Linuxboot](https://www.linuxboot.org/), [Ownerboot](https://sr.ht/~amjoseph/ownerboot/) and any similar payloads,
those are not legacy and they definitely have their places in the project, though, at the time of writing, no such payload is offered in NixOS.

In nixpkgs, legacy boot forces a dichotomy between `boot.loader.efi` and... at least two legacy bootloaders **in tree**: GRUB and a family of UBoot/extlinux-compatible/etc.

In the case of GRUB, there are increasing problems with this bootloader:

> - Upstream do not do releases anymore: https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/tools/misc/grub/default.nix#L62-L350

Reviewer 2: Broken link, because it references a branch instead of a commit.

---
feature: uefi_boot_only_for_supported_systems
start-date: 2023-06-16
author: Ryan Lahfa
co-authors: (find a buddy later to help out with the RFC)
shepherd-team: (names, to be nominated and accepted by RFC steering committee)
shepherd-leader: (name to be appointed by RFC steering committee)
related-issues: (will contain links to implementation PRs)
---

# Summary
[summary]: #summary

NixOS should drop support for legacy boot and assumes that
all systems **supported by UEFI** can be booted off UEFI.
For legacy boot systems, an UEFI-providing environment bootloader should be used to polyfill
for UEFI features.

# Motivation
[motivation]: #motivation

Legacy boot is defined by the BIOS Boot specification, written in 1996: https://www.scs.stanford.edu/nyu/04fa/lab/specsbbs101.pdf.
Nowadays, computers are defaulting to UEFI more and more for the extended features provided (e.g. SecureBoot, native network boot, etc.).

Nevertheless, many legacy boots machines or machines that does not have support for UEFI are used with NixOS: Single Board Computers for example
on other architectures.

> In nixpkgs, legacy boot forces a dichotomy between `boot.loader.efi` and... two legacy bootloaders: GRUB and a family of UBoot/extlinux-compatible/etc.

Reviewer: ```suggestion
In NixOS, legacy boot forces a dichotomy between `boot.loader.efi` and... two legacy bootloaders: GRUB and a family of UBoot/extlinux-compatible/etc.
```

BIOS also hold back all improvements in the boot story, considering we are not held back by this, is very interesting for NixOS.

# Detailed design
[design]: #detailed-design

Legacy boot is booted via UBoot or any bootloader able to provide an UEFI environment.

Then, we boot using any usual bootloader, e.g. systemd-boot.

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

Fedora is considering doing this: https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/GOETDM5SWINBX5ZDV37SWMHIPRRUVVTT/.

Asahi Linux is booting UEFI via UBoot without EFI variables and it has been great for them.

People who want to boot off strange partitions at boot-time can exploit UEFI drivers capability to load any filesystem driver and
open the EFI System Partition in ZFS if they really insist.

# Drawbacks
[drawbacks]: #drawbacks

- 2-stage boot for legacy BIOS systems is more than 1-stage boot
- Increased internal complexity in the boot story of NixOS
- Increased load and reliance on UBoot
- The runtime service `SetVariable` will probably stay highly unstable for a while (variable storage)

# Alternatives
[alternatives]: #alternatives

> - Keeping legacy BIOS and hoping someday it will be a bit more maintained

Reviewer 3: Keeping the option, (switching to just tracking the snapshots,) but change the defaults for new users?

Author: Yes, definitely

Author: (but this alternative could also be considered my proposal as in: I want to introduce long term deprecation then removal of this)

Reviewer 3: GRUB2 gives a lot of nice features beyond BIOS Boot, even features that I have had to use, though. 

On the other hand, making it non-default (not that everyone knows or wants to learn using those features…) could be done on a much shorter time horizon than deprecation discussions, I think

Author: Fair @ making it non default.

(But, to be completely fair, this will turn the GRUB2 question into a "maintenance in-tree" issue as currently, no one really is interested into this problem in nixpkgs.)

Reviewer 3: The problems you quote are mostly _NixOS_ side, and yeah, for anyone who understands how to make it work, it's the Nixpkgs side that is useful — and sufficient…

(Oh well, I just use GRUB2 UEFI from Nixpkgs while only using `modules/services` from NixOS)

Author: Well, the Nixpkgs side is not too beautiful due to how GRUB2 upstream development practices IMHO. It would be fine if people stepped in and did it like other distros (e.g. https://github.com/rhboot/grub2). This is not the case and there does not seem to be any appetite among current maintainers AFAIK.

Reviewer 3: I think that's an overkill, whatever snapshot passes our tests plus maybe a stolen patch from another downstream here and there might be enough

Author: Sure, what I am saying is no one is even doing the "stolen patch" except people who actually doesn't care that much about GRUB :).

Reviewer 3: I mean, if you don't have Secure Boot enabled anyway, GRUB just works on a ton of systems even without maintenance. Churn is a problem, not an end-goal positive, after all.

Author: This is simply not true: https://github.com/NixOS/nixpkgs/pull/235222 we already had boot broken (I don't remember if we had issues, but we had people coming to us on that subject in the Matrix channels) for our users recently because as I repeat it in the RFC (and this was cited in the original RFC text and still is.)

GRUB maintenance is not done, it is the default bootloader and our users do not see any real reason to change of bootloader.

I would not have any beef with this situation if GRUB was some optional bootloader, and we had replacements, but GRUB lack of maintenance is pushing churn on other maintainers who doesn't care at all about GRUB and has to do the work for it and has to consider it, if the answer to that is: "GRUB doesn't need maintenance anyway".

I am very inclined now to ensure we don't default for GRUB anymore for UEFI boot right now.

Reviewer 3: Oh right, `copyKernels` is not the default, that could lead to catches.
